taichi/backends/opengl/opengl_api.cpp

```diff
@@ -286,15 +305,17 @@ struct CompiledKernel {
 
   explicit CompiledKernel(const std::string &kernel_name_,
                           const std::string &kernel_source_code,
-                          int num_groups_,
+                          int num_groups_, RangeSizeEvaluator rse_,
```

Comment:
BTW, I think all the places that passes `RangeSizeEvaluator` can be made `const RangeSizeEvaluator&` (or you can move them)? Otherwise we are copying this object every time.

Reply:
To be clear, it is `using RangeSizeEvaluator = std::optional<std::function<...>>`, will `std::optional` copy it's content when itself copied? Anyway, I'll add `std::move` to prevent unnecessary copies.

Reply:
> will std::optional copy it's content when itself copied? 

Yes it will, see https://en.cppreference.com/w/cpp/utility/optional/optional

Reply:
nit: `move` in L312 as well?